### PR TITLE
perf(p2p): Reduce the p2p metrics overhead.

### LINKED ISSUE #2840: Lower allocation overhead of metrics in peer.Send
### Summary

In Memory profiles of full nodes, we see large amounts of heap allocation in p2p.send coming from Prometheus logging, that is ~entirely duplicate work. 
image

This is coming from these lines: https://github.com/cometbft/cometbft/blob/b8466398aacfc6f491ac54e741fce6750fa1cb9e/p2p/peer.go#L284-L292

We should be able to have the labels already cached (or at minimum the peer ID label cached) in the existing peer.Metrics instance.

### Problem Definition

Lower heap allocation and time overhead of sending something to peers. I have no idea how much this is a dominant term in wall clock time, though I do see recv time and send time routinely in CPU profiles.

### Proposal

Cache the peer ID label in p.metrics.PeerSendBytesTotal, so this is re-done on every packet send

### PR DESCRIPTION
**Pull request by ValarDragon**
Closes #2840

We do this, by the calls to Prometheus (which have many allocations) to be significantly batched, and importantly not blocking the send routine and recv routine.

This is a 25% speedup to recvRoutine, and 60% speedup to peer.Send. 

By napkin estimate, this also saves 8% of newObject call time, which hopefully also means over 8% of the GC overhead times

---

#### PR checklist

- [ ] Tests written/updated
- [ ] Changelog entry added in `.changelog` (we use [unclog](https://github.com/informalsystems/unclog) to manage our changelog)
- [ ] Updated relevant documentation (`docs/` or `spec/`) and code comments
- [ ] Title follows the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) spec

### CODE REVIEW

**Comment by ValarDragon** [comment]
This was a 60% reduction to GC time :eyes: from one benchmark, so this actually sped up the net CPU profile time by 30% for Osmosis based on two 1 hour profiles. 

(Note Osmosis GC has many other terms eliminated, e.g. debug logs, and my other open PR's. On main Comet branches, GC will be comet debug log dominated)

**Comment by cason** [comment]
How did you get this performance improvement numbers? Because they are impressive.

I probably need a way to test connections in isolation to experimentally check the impact of this changes.

**Comment by cason** [comment]
A second point to made, that worries from a while, is the fact that metric caches are private types. This is not exactly a problem, as they are internal to this package, but the problem is that they are provided in some constructors (e.g., `peerConfig`), rendering impossible to use Public types in the `p2p` package because it is private.

In summary, is there any way to render the metrics cache types part of the public `p2p.Metrics` type?

**Review by cason — COMMENTED** [review]
Some general comments regarding the implementation.

For me the only blocker is that we are updating these stats every 10 seconds, instead of every send/receive call. Should we reconsider this granularity?

**Comment by melekes** [comment]
A more general comment is: how confident are we that we need these metrics? If they are costing us so much CPU.

File: p2p/peer.go [hunk 1]
```diff
@@ -380,6 +378,27 @@ func (p *peer) metricsReporter() {
 			}
 
 			p.metrics.PeerPendingSendBytes.With("peer_id", string(p.ID())).Set(sendQueueSize)
+			// Report per peer, per message total bytes, since the last interval
+			func() {
```
[the remark this answers is not in the document]
Comment (ValarDragon) [reply p2p/peer.go:382]: Good catch, didnt realizd it was rhis long. 

I think 10s is fine, at least I've never looked at these in finer grained views. We could perhaps do 1s or 5s, but i don't think this is a big issue

File: p2p/metrics.go [hunk 2]
```diff
@@ -41,12 +41,70 @@ type metricsLabelCache struct {
...
+
+func getMsgType(i any) reflect.Type {
+	return reflect.TypeOf(i)
+}
+
 // ValueToMetricLabel is a method that is used to produce a prometheus label value of the golang
 // type that is passed in.
 // This method uses a map on the Metrics struct so that each label name only needs
 // to be produced once to prevent expensive string operations.
 func (m *metricsLabelCache) ValueToMetricLabel(i any) string {
```
[the remark this answers is not in the document]
Comment (ValarDragon) [reply p2p/metrics.go:105]: Oh haha, ok will make a follow up api breaking pr then

**Comment by ValarDragon** [comment]
I actually found them really helpful for debugging bandwidth. The metric is still a cost after this pr, but I need to reprofile after casons point that I'm copying the struct here

File: p2p/metrics.go [hunk 3]
```diff
@@ -41,12 +41,69 @@ type metricsLabelCache struct {
...
+		}
+	}
+}
+
+func (c *peerPendingMetricsCache) AddPendingRecvBytes(msgType reflect.Type, addBytes int) {
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+	if entry, ok := c.perMessageCache[msgType]; ok {
+		entry.pendingRecvBytes += addBytes
+		c.perMessageCache[msgType] = entry
```
Comment (melekes) [inline p2p/metrics.go:88]:
```suggestion
		c.perMessageCache[msgType] = entry
```

https://github.com/cometbft/cometbft/pull/3411/files#r1666530037

File: p2p/metrics.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (melekes) [reply p2p/metrics.go:105]: It's not breaking since `metricsLabelCache` was never exported, no?

**Comment by melekes** [comment]
@ValarDragon I've removed `metricsLabelCache` because I've realized it's now unused. I.e., **metric labels are not shared between peer caches** because you've removed the `ValueToMetricLabel` call from `peer`. Could you get the profile with the latest changes and confirm that the overhead is gone? Thanks 🙏

File: p2p/metrics.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (cason) [reply p2p/metrics.go]: It is private, right?

**Review by cason — APPROVED** [review]
Great! I loved removing this `mlc` private parameter for `Peer` and other (public) components.

The only part I don't agree is to report metrics only every 10s:

https://github.com/cometbft/cometbft/blob/f503f25cfc985543fad97bcde9b6146ee424f078/p2p/peer.go#L20

While now we report then on every send/receive method.

Should we reduce it to, say, 1s, 0.5s?

**Comment by melekes** [comment]
> The only part I don't agree is to report metrics only every 10s:

why? Prometheus default scrape interval is 1m https://prometheus.io/docs/prometheus/latest/configuration/configuration/

**Comment by cason** [comment]
> > The only part I don't agree is to report metrics only every 10s:
> 
> why? Prometheus default scrape interval is 1m https://prometheus.io/docs/prometheus/latest/configuration/configuration/

If you say so, great. Lets merge it.

Then, can we somehow compare the Prometheus output with and without this PR?

**Comment by ValarDragon** [comment]
I can try to see if I notice any difference in our prod rpc's Grafana metrics. But as @melekes noted, its not that fine grained, so not sure I'll notice anything

**Comment by cason** [comment]
I'll wait for @melekes be back, but I am good with the current state of this PR.

**Comment by melekes** [comment]
> > > The only part I don't agree is to report metrics only every 10s:
> > 
> > 
> > why? Prometheus default scrape interval is 1m https://prometheus.io/docs/prometheus/latest/configuration/configuration/
> 
> If you say so, great. Lets merge it.
> 
> Then, can we somehow compare the Prometheus output with and without this PR?

Please ignore my comment. We lose granularity if the scrape interval is 10s >= Prometheus scrape interval (10s).   Let's lower it down to 1s. We are spawning the goroutine now, so it shouldn't be a burden.

File: p2p/test_util.go
Comment (cason) [inline p2p/test_util.go:1]: We have to remove line 44 added by another PR.

**Pull request merged**